File: src/chapters/differential-equations/content/code/modified-euler.cpp
#include <iostream>
#include <functional>
using namespace std;

// Modified Euler (midpoint RK2, predictor-corrector).
double modified_euler(function<double(double, double)> f, double t0, double y0, double h, int n) {
    double t = t0, y = y0;
    for (int i = 0; i < n; ++i) {
        double k1 = f(t, y), k2 = f(t + h / 2, y + h / 2 * k1);
        y += h * k2; t += h;
    }
    return y;
}

int main() {
    auto f = [](double t, double y) { return 2 * y - 10 * t * t + 2 * t; };
    cout << "y(1) = " << modified_euler(f, 0.0, 1.0, 0.1, 10) << "\n";
}

File: src/chapters/differential-equations/content/code/modified-euler.r
# Modified Euler (midpoint RK2, predictor-corrector).
modified_euler <- function(f, t0, y0, h, n) {
  t <- t0
  y <- y0
  for (i in 1:n) {
    k1 <- f(t, y)
    k2 <- f(t + h / 2, y + h / 2 * k1)   # slope at the midpoint
    y <- y + h * k2
    t <- t + h
  }
  list(t = t, y = y)
}

f <- function(t, y) 2 * y - 10 * t^2 + 2 * t
cat("y(1) =", modified_euler(f, 0.0, 1.0, 0.1, 10)$y, "\n")
